content/blog/apl/2021-02-28_look_at_hpkg.md

```diff
@@ -0,0 +1,115 @@
++++
+type = "blog"
+title = "A Brief Look Inside HPKG"
+author = "apl"
+date = "2021-02-28 12:00:00+12:00"
+tags = ["haiku", "haikudepot", "hpkg", "hpkr", "format"]
++++
+
+The Haiku operating system has a packaging system that leverages a clever file format called HPKG.  This article provides a simplistic overview of how the file format is structured.
+
+## Installing and Uninstalling HPKGs
+
+To get a package installed on their Haiku computer, a user would download a package file in a format called HPKG.  The user would typically do this via the [HaikuDepot](https://www.haiku-os.org/docs/userguide/en/applications/haikudepot.html) desktop application or by using the `pkgman` command line tool.  Once an HPKG file is downloaded and moved into place at `/system/packages`, the contents of the package appear read-only in the file system.  An example file might be `pe-2.4.5-8-x86_64.hpkg` which would provide the necessary files for the popular Pe text editor to be used.
+
+At some later date, should the user no longer want the package installed, they would move it out of `/system/packages` and instantly the files of the package would vanish from the file system.  The user would typically use HaikuDepot or `pkgman` to uninstall the package, but it would work equally well to just use [Tracker](https://www.haiku-os.org/docs/userguide/en/tracker.html) to remove the HPKG because it is simply a file.
+
+The HPKG acts as an archive of the package's files but instead of the Haiku operating system unpacking the files into its own storage, it is instead mounting the HPKG file contents into its own file system.
```

Comment:
```suggestion
The HPKG acts as an archive of the package's files but instead of the Haiku operating system unpacking the files into its own storage, it is instead mounting the HPKG file contents into its own virtual file system.
```

Reply:
Overall I think this post is great but I think specifying that the package file system is virtual might make things more clear.

Reply:
Thanks; a good suggestion - I have committed that.